docs/man/sddf/en/sddf.1.md
% SDDF(1)
% Naohiro CHIKAMATSU <n.chika156@gmail.com>
% December 2021

# NAME

serial –  Search & Delete Duplicatetd File

# SYNOPSIS

**sddf** [OPTIONS] PATH

# DESCRIPTION
**sddf** looks for duplicate files under the specified directory and  
creates a list of them (default: duplicated-file.sddf).  
If the list is executed with sddf arguments, sddf delete the file based  
on the contents of the list.

# EXAMPLES
**Search for duplicate files under the current directory**  

    $ sddf .  

**Remove duplicate files**  

    $  sddf duplicated-file.sddf  

# OPTIONS
**-o, **--output**
:   Specify the file name of the duplicate file list.  

**-h**, **--help**
:   Show help message.

**-v**, **--version**
:   Show sddf command version.

# EXIT VALUES
**0**
:   Success

**1**
:   Error when specifying the argument of the sddf command, or error during file operation

# BUGS
See GitHub Issues: https://github.com/nao1215/mimixbox/issues

# LICENSE
The MimixBox project is licensed under the terms of the MIT license and Apache License 2.0.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.28.07] - 2021-12-12
 - unix2dos command.
   - Changed to print the message being converted.
## [0.28.06] - 2021-12-12
### Changed
 - wc command.
   - Fixed the bug that did not count the number of rows of data passed from PIPE correctly.
 - cat / md5sum / sha1sum / sha256sum / sha512sum command.
   - Fixed a bug that the way pipes are handled changes depending on whether there is an option or not.
 - rm command.
   - Changed to receive data from pipe.
   - When removing multiple files, processing continues even if remove fails in the middle.
 - sddf command.
   - Changed to get the file path as much as possible without stopping the process even if an error occurs.
   - Changed to output "." continuously while getting the file path.
   - Changed to show the size of the deleted files.
   - Important files under /dev and /boot and etc. are excluded from deletion.
   - Speeded up checksum calculation with goroutine.
   - Removed named PIPE from checksum calculation. 
     - The checksum calculation for the named PIPE will stop unless there is writing to the named PIPE. It's looks like deadlock. To avoid this problem, exclude named PIPE from target file list.
 - basename command.
   - Print error message if there is no argument.
   - Matched the result with Coreutils when the user specified an empty string.
   - Matched the result with Coreutils when the user specified multiple arguments.
 - mkfifo command.
   - Changed to continue processing even if multiple named pipes fail while creating.
   - Changed to print the path of the file that failed to be created on error.
 - dos2unix command.
   - Changed to print the message being converted.
## [0.28.00] - 2021-12-08
### Added
 - sddf command. Search & Delete Duplicated Files.
### Changed
 - mimixbox command.
   - When an error occurs, the name of the applet that caused the error is displayed.
 - wc command.
   - Fixed the bug that the file specified by the argument is not referenced when the wc command is connected by pipe.
   - Display the count result even if the argument is directory.
   - When an error occurs, wc command display its name.
 - cat command. 
   - Fixed the bug that the cat command does not refer to the argument when pipe is used.
   - Fixed the bug that some lines do not have line numbers when the --number option is specified and there is a line feed code in the file-to-file concatenation.
   - Concatenate here documents and files.
 - mkdir command.
   - Output an error instead of a help message when no argument is specified.
 - touch command.
   - Continue processing even if an error occurs
 - cp command.
   - Print error if the -r option is not attached when copying the directory.
   - Print error if the copy destination is in the hierarchy below the copy source directory.
   - Fixed the bug that the directory hierarchy of the source path is copied as it is when copying the directory. The correct process is to copy from the base name of the source path.
 - md5sum / sha1sum / sha256sum / sha512sum command.
   - Print the error if the argument is directory or non-existent file.
   - If PIPE and file path are passed at the same time, the PIPE data will be ignored.
 - which command.
   - Change exit-status frmo succes to error if which command can't find binary.
   - Changed the specification that allows only one command to be searched.
 - nl command.
   - Fixed the bug that the cat command does not refer to the argument when pipe is used.
   - Delete unused line feed.
   - Fixed a bug that line numbers do not match when concatenating PIPE data and files.
   - Print the error when specifying the file that does not exist.
## [0.27.10] - 2021-12-01
### Added
 - Add ShellSpec tetsing framework for integration test.
 - ut(Unit Test)／it（Integration Test）target in Makefile.
 - full-install／remove target in Makefile.
### Changed
 - MimixBox command.
   - Display help message when --install, --full-install, --remove are specified and there is no optional argument.
   - Make the error if the directory specified by the user does not exist when executing --install, --full-install, --remove.
   - Fixed the bug that always determines that the applet does not exist when the user specifies the applet with the full path.
 - Makefile.
   - Display accurate coverage by specifying "-coverpkg=./..." in the unit test.
 - Commands that read file (dos2unix, expand, head, tail, unexpand, wc)
   - Fixed the bug that caused Runtime Error when reading the empty file.
 - mkdir command.
   - Create multiple directories with a single command.
     Previously, an error occurred when specifying multiple directories.
 - cp command.
   - Fixed the bug that files cannot be copied when the copy destination is only the directory name.
   - Fixed the bug that the cp command could not copy the directory with complex tree structure.
 - wc command.
   - Unified output format with Coreutils.
   - Fixed the bug that the -L option was not implemented.
 - unix2dos command.
   - Fixed the bug that print incorrect command name. 

## [0.27.1] - 2021-11-29
### Added
 - sl commad.
 - wc command.
### Changed
- expand environment variables in PATH specified by argument
  (mimixbox, cp, ln, mkdir, mkfifo, mv, rm, rmdir, touch, fakemovie, chroot,
   cat, dos2unix, expand, head, tac, tail, unexpand, unix2dos)
- Fixed the bug that was using the wrong installation path.
## [0.25.1] - 2021-11-27
### Added
 - hostid commad(Does not work properly)
 - md5sum command.
 - seq command.
 - sha1sum／sha256sum／sha512sum command
### Changed
- error output destination of MimixBox from STDOUT to STDERR
- all commands to support redirects.
## [0.22.0] - 2021-11-25
### Added
 - dos2unix/unix2dos command.
 - expand/unexpand command.

internal/applets/shellutils/sddf/sddf.go
//
// mimixbox/internal/applets/shellutils/sddf/sddf.go
//
// Copyright 2021 Naohiro CHIKAMATSU
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package sddf

import (
	"crypto/md5"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/cheggaaa/pb/v3"
	"github.com/jessevdk/go-flags"
	mb "github.com/nao1215/mimixbox/internal/lib"
)

const cmdName string = "sddf"
const ext string = ".sddf"

type Paths []string

type fileInfo struct {
	path     string
	checksum string
}

const version = "1.0.2"

var osExit = os.Exit

// Exit code
const (
	ExitSuccess int = iota // 0
	ExitFailuer
)

type options struct {
	Output  string `short:"o" long:"output" default:"duplicated-file.sddf" description:"Change output file-name without extension"`
	Version bool   `short:"v" long:"version" description:"Show sddf command version"`
}

func Run() (int, error) {
	var opts options
	var args []string
	var err error

	if args, err = parseArgs(&opts); err != nil {
		return ExitFailuer, nil
	}
	return sddfMainSeq(os.ExpandEnv(args[0]), opts)
}

func sddfMainSeq(path string, opts options) (int, error) {
	if !mb.Exists(path) {
		return ExitFailuer, errors.New(path + " does not exists")
	}

	if mb.IsFile(path) {
		return restoreAndDelete(path)
	}
	return search(path, opts)
}

func restoreAndDelete(path string) (int, error) {
	if !strings.HasSuffix(path, ext) {
		return ExitFailuer, errors.New(path + ": file format is not *.sddf")
	}

	df, err := restore(path)
	if err != nil {
		return ExitFailuer, err
	}

	return deleteFiles(df)
}

func deleteFiles(df map[string]Paths) (int, error) {
	var status int = ExitSuccess
	var deleteFileList []string

	fmt.Fprintln(os.Stdout, "Decide delete target files")
	for _, v := range df {
		list, err := decideDeleteTarget(v)
		if err != nil {
			return ExitFailuer, err
		}
		deleteFileList = append(deleteFileList, list...)
	}

	fmt.Fprintln(os.Stdout, "Start deleting files")
	var sumByteSize int64 = 0
	for _, v := range deleteFileList {
		size, err := mb.Size(v)
		if err != nil {
			status = ExitFailuer
			fmt.Fprintln(os.Stdout, "Delete(Failuer): "+v)
			continue
		}

		err = mb.RemoveFile(v, false)
		if err != nil {
			status = ExitFailuer
			fmt.Fprintln(os.Stdout, "Delete(Failuer): "+v)
		} else {
			fmt.Fprintln(os.Stdout, "Delete(Success): "+v+": "+strconv.FormatInt(size, 10)+"Byte")
		}
		sumByteSize += size
	}
	fmt.Fprintln(os.Stdout, "End deleting files. Size="+strconv.FormatInt(sumByteSize, 10)+"Byte")
	return status, nil
}

func decideDeleteTarget(paths []string) ([]string, error) {
	fileModTime := map[string]int64{}
	for _, path := range paths {
		f, err := os.Open(path)
		if err != nil {
			return nil, err
		}
		defer f.Close()

		finfo, _ := f.Stat()
		dsec := finfo.ModTime().Unix()
		fileModTime[path] = dsec
	}

	var newest int64 = 0
	var leaveKey string = ""
	for k, v := range fileModTime {
		if newest < v {
			newest = v
			leaveKey = k
		}
	}

	deleteTarget := []string{}
	for k := range fileModTime {
		if k == leaveKey {
			continue
		}
		deleteTarget = append(deleteTarget, k)
	}
	return deleteTarget, nil
}

func processing(cancel chan struct{}) {
	i := 0
	for {
		select {
		case <-cancel:
			time.Sleep(1 * time.Second)
			return
		default:
			if i != 80 {
				fmt.Fprintf(os.Stdout, ".")
				time.Sleep(100 * time.Millisecond)
				i++
			} else {
				fmt.Fprintln(os.Stdout, "")
				i = 0
			}
		}
	}
}

func restore(path string) (map[string]Paths, error) {
	df := map[string]Paths{}
	lines, err := mb.ReadFileToStrList(path)
	if err != nil {
		return nil, err
	}

	fmt.Fprintln(os.Stdout, "Restore data from "+path)
	bar := pb.Simple.Start(len(lines))
	bar.SetMaxWidth(80)
	var checksum string
	var paths = Paths{}
	lines = mb.ChopAll(lines)
	for _, v := range lines {
		bar.Increment()
		if isChecksumLine(v) {
			checksum = v
			paths = Paths{}
			continue
		}
		if v == "" {
			df[checksum] = paths
			checksum = ""
			paths = Paths{}
			continue
		}
		if v != "" && !isChecksumLine(v) {
			paths = append(paths, v)
		}
	}
	bar.Finish()

	return df, nil
}

func isChecksumLine(line string) bool {
	return strings.HasPrefix(line, "[") && strings.HasSuffix(line, "]") && len(line) == 34
}

func search(path string, opts options) (int, error) {
	files := findFiles(path)

	if len(files) == 0 {
		fmt.Fprintln(os.Stdout, path+" has no file")
		return ExitSuccess, nil
	}
	return dumpToFile(getSameFiles(files), decideOutputFileName(opts.Output))
}

func findFiles(path string) []string {
	fmt.Fprintln(os.Stdout, "Get all file path at "+path)

	cancel := (make(chan struct{}))
	go processing(cancel)
	_, files, _ := mb.Walk(path, true)
	files = excludeImportantFiles(files)
	close(cancel)
	fmt.Fprintln(os.Stdout)

	return files
}

func excludeImportantFiles(files []string) []string {
	ng := []string{"/boot", "/dev", "/etc", "/lib", "/lib32", "/lib64", "/libx32", "/lost+found",
		"/proc", "/root", "/run", "/sys", "/bin", "/sbin"}

	newFileList := []string{}
	for _, path := range files {
		for i, v := range ng {
			if strings.HasPrefix(path, v) {
				break
			}
			if (i + 1) == len(ng) {
				newFileList = append(newFileList, path)
			}
		}
	}
	return newFileList
}

func dumpToFile(df map[string]Paths, output string) (int, error) {
	f, err := os.Create(output)
	if err != nil {
		return ExitFailuer, err
	}
	defer f.Close()

	data := ""
	bar := pb.Simple.Start(len(df))
	bar.SetMaxWidth(80)
	for k, v := range df {
		data += "[" + k + "]\n"
		for _, path := range v {
			data += path + "\n"
		}
		data += "\n"
		bar.Increment()
	}
	bar.Finish()

	b := []byte(data)
	_, err = f.Write(b)
	if err != nil {
		return ExitFailuer, err
	}

	fmt.Fprintln(os.Stdout, "See duplicated file list: "+output)
	fmt.Fprintln(os.Stdout, "If you delete files, execute the following command.")
	fmt.Fprintln(os.Stdout, "$ sddf "+output)
	return ExitSuccess, nil
}

func getSameFiles(files []string) map[string]Paths {
	df := calcChecksum(files)
	for k, v := range df {
		if len(v) <= 1 {
			delete(df, k)
		}
	}
	return df
}

func calcChecksum(files []string) map[string]Paths {
	totalFileNum := len(files)
	df := map[string]Paths{}

	fmt.Fprintln(os.Stdout, "Find the same file on a file content basis")

	ch := make(chan fileInfo, totalFileNum)
	if totalFileNum < 100 {
		go calcChecksumThread(files, ch)
	} else {
		for n := 0; n < 10; n++ {
			ratio := totalFileNum / 10
			first := n * ratio
			last := (n + 1) * ratio
			if n == 9 {
				last = totalFileNum
			}
			go calcChecksumThread(files[first:last], ch)
		}
	}
	bar := pb.Simple.Start(totalFileNum)
	bar.SetMaxWidth(80)
	for n := 0; n < totalFileNum; n++ {
		fi := <-ch
		paths, ok := df[fi.checksum]
		if ok {
			paths = append(paths, fi.path)
			df[fi.checksum] = paths
		} else {
			df[fi.checksum] = []string{fi.path}
		}
		bar.Increment()
	}
	close(ch)
	bar.Finish()
	return df
}

func calcChecksumThread(files []string, ch chan fileInfo) {
	for _, v := range files {
		var fi fileInfo = fileInfo{"", ""}
		absPath, err := filepath.Abs(v)
		if err != nil {
			//fmt.Fprintln(os.Stderr, cmdName+": "+err.Error())
			ch <- fi
			continue
		}

		checksum, err := mb.CalcChecksum(md5.New(), v)
		if err != nil {
			//fmt.Fprintln(os.Stderr, cmdName+": "+err.Error())
			ch <- fi
			continue
		}
		fi.path = absPath
		fi.checksum = checksum
		ch <- fi
	}
}

func decideOutputFileName(output string) string {
	if strings.HasSuffix(output, ext) {
		return output
	}
	return output + ext // Do not exclude original file-extension
}

func parseArgs(opts *options) ([]string, error) {
	p := initParser(opts)

	args, err := p.Parse()
	if err != nil {
		return nil, err
	}

	if opts.Version {
		mb.ShowVersion(cmdName, version)
		osExit(ExitSuccess)
	}

	if !isValidArg(args, *opts) {
		showHelp(p)
		osExit(ExitFailuer)
	}

	return args, nil
}

func isValidArg(args []string, opts options) bool {
	return len(args) == 1
}

func showHelp(p *flags.Parser) {
	p.WriteHelp(os.Stdout)
}

func initParser(opts *options) *flags.Parser {
	parser := flags.NewParser(opts, flags.Default)
	parser.Name = cmdName
	parser.Usage = "[OPTIONS] PATH"

	return parser
}
